### PR TITLE
:seedling: Support for ClusterClass template

### LINKED ISSUE #1267: Implement the ClusterClass API
**User Story**

As an operator, I would like to be able to use the [ClusterClass](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20210526-cluster-class-and-managed-topologies.md) API for all the reasons listed in the [Motivations paragraph](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20210526-cluster-class-and-managed-topologies.md#motivation) of said proposal. 

**Detailed Description**

As I see it, the only missing part is the addition of a `Metal3ClusterTemplate` CRD, which embeds the `Metal3ClusterSpec` type. If I understand it correctly, no additional controller logic is required, as it is handled by the ClusterAPI itself :
- https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/topology/cluster/cluster_controller.go#L275
- https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/topology/cluster/desired_state.go#L57
- https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/topology/cluster/desired_state.go#L160
- https://github.com/kubernetes-sigs/cluster-api/blob/main/controllers/external/util.go#L197

**Anything else you would like to add:**

It seems this is a quick-win feature, which was already asked for on the community Slack :
- https://kubernetes.slack.com/archives/CHD49TLE7/p1684166116481899
- https://kubernetes.slack.com/archives/CHD49TLE7/p1636058124100500

/kind feature

### PR DESCRIPTION
**Pull request by p-strusiewiczsurmacki-mobica**
**What this PR does / why we need it**:

This PR adds support for ClusterClass API as discussed in #1267 

**Which issue(s) this PR fixes**:
Fixes #1267

### CODE REVIEW

**Comment by Rozzii** [comment]
/ cc @kashifest

File: examples/clusterclass/cluster/cluster.yaml
Comment (lentzi90) [inline examples/clusterclass/cluster/cluster.yaml:107]: The cluster-template should be separate from the ClusterClass IMO, and we should not rely on the `CLUSTER_NAME` in the ClusterClass at all. That way it will be possible to reuse the ClusterClass for multiple clusters.

File: examples/generate.sh
Comment (lentzi90) [inline examples/generate.sh:101]: Why is this needed?

File: api/v1beta1/metal3clustertemplate_webhook_test.go
Comment (lentzi90) [inline api/v1beta1/metal3clustertemplate_webhook_test.go:37]: Could you add one test for when the template is invalid also?

**Review by lentzi90 — COMMENTED** [review]
Thanks for working on this! I think the code is looking good already but I'm a bit confused about the kustomizations under examples...
Why is it split up in cluster, controlplane and machinedeployment? I would expect to have one kustomization with everything needed for the ClusterClass (Metal3MachineTemplates, Metal3ClusterTemplate, KubeadmControlPlaneTemplate, KubeadmConfigTemplates, and also Metal3DataTemplates and IPPools). Then the ClusterClass kustomization would be "self-contained" and we can have a separate cluster-template to create clusters based on this class. What do you think?

File: examples/clusterclass/cluster/cluster.yaml
Comment (p-strusiewiczsurmacki-mobica) [reply examples/clusterclass/cluster/cluster.yaml]: Changed that. I previously left the old names as it was easier for me to make clusterclass example similar to the non-clusterclass one as possible.
Can you check if this is what you've meant?

File: examples/generate.sh
Comment (p-strusiewiczsurmacki-mobica) [reply examples/generate.sh]: Mainly `--reorder=none` is required for clusterclass example. By default kustomization will use `reorder=legacy` and will reorde definitions in the outcome file alphabetically. Now, if the definitions are ordered and ClusterClass will be in the outcome file **above** the Metal3 definitions (e.g. `Metal3DataTemplate`), during deletion with `make delete-examples-clusterclass` the operation will fail as k8s will first delete ClusterClass and associated Metal3 objects and will try to delete Metal3 objects once again. At least that's what's happening to me, hence `reored=none` was added for ClusterClass example and `reorder=legacy` was added for non-ClusterClass example generation.

However, I've deleted this `reorder=legacy` and now only `reorder=none` will be used and only for ClusterClass.

File: api/v1beta1/metal3clustertemplate_webhook_test.go
Comment (p-strusiewiczsurmacki-mobica) [reply api/v1beta1/metal3clustertemplate_webhook_test.go:37]: Done :)

**Comment by p-strusiewiczsurmacki-mobica** [comment]
Added changes required by @lentzi90 

Additionally added some changes in the Tiltfile and other scripts.
Now, to test with Tilt below commands should be sufficient:
```
make tilt-settings-clusterclass
make tilt-up
make generate-examples-clusterclass
make deploy-examples-clusterclass
```

`make titl-settings-clusterclass` will now generate settings that should enable ClusterTopology in Cluster API.

Additionally, I was able to change the example available in [metal3-dev-env](https://github.com/metal3-io/metal3-dev-env) and to deploy and provision cluster using ClusterClass, so I'm moving this from `draft`.

If anyone wants to try to test it with `metal3-dev-env` [here's my branch with the changes.](https://github.com/p-strusiewiczsurmacki-mobica/metal3-dev-env/tree/clusterclass-support). Cluster should be provisioned if `make test` will be executed, but test itself will fail as I don't know how to enable `ClusterTopology` in the 'internal' cluster yet.

File: examples/generate.sh
Comment (lentzi90) [inline examples/generate.sh:1]: This way of generating manifests with `kustomize` and `envsubst` doesn't seem to really give us anything useful from what I can see. I'm fine having it here is if it useful to you but I wonder if we should not just put the ClusterClass and cluster-template in `templates` instead? No kustomize needed (we are anyway not doing anything useful with it) and they would be easily discovered and useful. It also matches exactly what [CAPI has for CAPD](https://github.com/kubernetes-sigs/cluster-api/tree/main/test/infrastructure/docker/templates).

**Review by lentzi90 — COMMENTED** [review]
I would not worry about the tilt setup here honestly. If you want to work on it, fine, but isn't it easier to just use CAPI's directly? 
The template and ClusterClass that you have here will work great with that already. The only thing you have to do is to add the folder to `template_dirs` in the tilt settings.

File: examples/clusterclass/class/class.yaml
Comment (p-strusiewiczsurmacki-mobica) [reply examples/clusterclass/class/class.yaml]: Done. Added variables and patches to `ClusterClass` and referenced those from `Cluster` itself. Seems to work fine regarding to my tests.
Just one question - currently `Metal3ClusterTemplate` is [validated by the webhook for `controlPlaneEndpoint` definition](https://github.com/p-strusiewiczsurmacki-mobica/cluster-api-provider-metal3/blob/clusterclass-support/api/v1beta1/metal3clustertemplate_webhook.go#L65), meaning `something` has always be defined there. Should this validation be omitted and we should allow empty definition (without specifying controlPlaneEndpoint`)?

File: examples/clusterclass/class/class.yaml
Comment (p-strusiewiczsurmacki-mobica) [reply examples/clusterclass/class/class.yaml]: Done

File: hack/gen_tilt_settings.sh
Comment (p-strusiewiczsurmacki-mobica) [reply hack/gen_tilt_settings.sh]: Done

File: examples/generate.sh
Comment (p-strusiewiczsurmacki-mobica) [reply examples/generate.sh:1]: I've deleted customization files and moved `clusterclass.yaml` and `cluster.yaml` to `examples/templates/`.

File: examples/templates/clusterclass.yaml
Comment (lentzi90) [inline examples/templates/clusterclass.yaml:25]: What is this cloud provider?

File: Makefile
Comment (lentzi90) [inline Makefile:415]:
```suggestion
	CLUSTER_TOPOLOGY=true ./examples/generate.sh
```

File: examples/generate.sh
Comment (lentzi90) [inline examples/generate.sh:101]:
```suggestion
if [ -n "${CLUSTER_TOPOLOGY:-}" ]; then
```

File: examples/generate.sh
Comment (lentzi90) [inline examples/generate.sh:60]:
```suggestion
```

File: examples/templates/clusterclass.yaml
Comment (lentzi90) [inline examples/templates/clusterclass.yaml:63]: Too bad we have the clusterName on the Metal3DataTemplate and IPPool :disappointed: 
I guess this makes it quite hard to use them from the ClusterClass? Would it make more sense to have them separate or in the cluster.yaml?

File: examples/clusterclass/class/class.yaml
Comment (lentzi90) [reply examples/clusterclass/class/class.yaml]: Not sure I understood the question. I think the controlPlaneEndpoint is always required. How would the cluster work without it?

**Comment by furkatgofurov7** [comment]
/cc

Not to forget about this, I'll take a closer look by EOD

File: examples/generate.sh
Comment (furkatgofurov7) [inline examples/generate.sh:107]:
```suggestion
  # Generate cluster resources.
  "$ENVSUBST" -i "${SOURCE_DIR}/templates/cluster.yaml" >"${CLUSTER_GENERATED_FILE}"
  echo "Generated ${CLUSTER_GENERATED_FILE}"
```

File: Makefile
Comment (furkatgofurov7) [inline Makefile:520]: Don't you need to pass `|| true ` to all of these as in `delete-examples` above?

**Review by furkatgofurov7 — COMMENTED** [review]
First pass, looks good so far, thanks for the PR. Left some comments and questions, and also it would be nice to have a clearly described documentation on how to enable/use CC for cluster creation from the user's point of view. 

What we usually do was: add a documentation in this repo and a user-guide in https://github.com/metal3-io/metal3-docs/tree/main/docs/user-guide/src

File: examples/clusterclass/class/class.yaml
Comment (p-strusiewiczsurmacki-mobica) [reply examples/clusterclass/class/class.yaml]: Currently `Metal3ClusterTemplate` is required to have `controlPlaneEndpoint` defined, e.g:

```
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: Metal3ClusterTemplate
metadata:
  name: example-cluster-template
spec:
  template:
    spec:
      controlPlaneEndpoint:
        host: 127.0.0.1
        port: 6443
      noCloudProvider: true
```

But I think that makes it not reusable, right? Do we want to accept something like this (with no `controlPlaneEndpoint`):

```
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: Metal3ClusterTemplate
metadata:
  name: example-cluster-template
spec:
  template:
    spec:
      noCloudProvider: true
```

And then controlPlaneEndpoint should be defined in `Cluster` resource, I think?

File: examples/templates/clusterclass.yaml
Comment (p-strusiewiczsurmacki-mobica) [reply examples/templates/clusterclass.yaml:25]: TBH, I've copied it from the existing example: https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/examples/controlplane/controlplane.yaml#L28

File: Makefile
Comment (p-strusiewiczsurmacki-mobica) [reply Makefile]: Had no issues without those yet, but added those just in case.

File: examples/clusterclass/class/class.yaml
Comment (lentzi90) [reply examples/clusterclass/class/class.yaml]: Ok now I understand!
I think the way to handle this is to use variables in the ClusterClass. So set it like you already did in the Metal3ClusterTemplate, but then add a variable and patch to override it. Then it would be possible to create clusters like

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: my-cluster
spec:
  ...
  topology:
    class: example-clusterclass
    version: v1.29.0
    variables:
    - name: controlPlaneEndpointHost
      value: 192.168.0.100
```

File: examples/templates/clusterclass.yaml
Comment (lentzi90) [reply examples/templates/clusterclass.yaml:25]: Not sure about what this does honestly. I think we can remove it. We don't use it in the tests at least: https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/test/e2e/data/infrastructure-metal3/bases/cluster/cluster-with-kcp.yaml

**Review by lentzi90 — COMMENTED** [review]
I think this is almost ready to merge. I just have two things I would like to get resolved first.

1. Make variables for the controlPlaneEndpoint so it can easily be overridden when creating a Cluster from the ClusterClass.
2. Use the builtin variable (`builtin.cluster.name`) to set the cluster name on the IPPool and Metal3DataTemplate.

**Comment by p-strusiewiczsurmacki-mobica** [comment]
>     1. Make variables for the controlPlaneEndpoint so it can easily be overridden when creating a Cluster from the ClusterClass.
It's already there, unless you've meant something different: https://github.com/p-strusiewiczsurmacki-mobica/cluster-api-provider-metal3/blob/c6aa980e7fb91733078aa9d5e34388fcc3adcbf1/examples/templates/clusterclass.yaml#L342

>     2. Use the builtin variable (`builtin.cluster.name`) to set the cluster name on the IPPool and Metal3DataTemplate.

I am not sure if that's possible. [According to the docs](https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-class/write-clusterclass#builtin-variables)
```
builtin variables can be referenced in patches:
```

[It seems that ClusterClass patches can only match ControlPlane, InfrastructureCluster, MachineDeploymentClass and MachinePoolClass](https://pkg.go.dev/sigs.k8s.io/cluster-api/api/v1beta1#PatchSelectorMatch) so I don't see any way to match and patch `Metal3DataTemplate` or `IPPool`. Or am I missing something?

File: examples/templates/clusterclass.yaml
Comment (p-strusiewiczsurmacki-mobica) [reply examples/templates/clusterclass.yaml]: OK, removed all references to `cloud-provider: baremetal`.

**Comment by lentzi90** [comment]
> It's already there, unless you've meant something different: https://github.com/p-strusiewiczsurmacki-mobica/cluster-api-provider-metal3/blob/c6aa980e7fb91733078aa9d5e34388fcc3adcbf1/examples/templates/clusterclass.yaml#L342

Excellent, not sure how I missed it!

> [It seems that ClusterClass patches can only match ControlPlane, InfrastructureCluster, MachineDeploymentClass and MachinePoolClass](https://pkg.go.dev/sigs.k8s.io/cluster-api/api/v1beta1#PatchSelectorMatch) so I don't see any way to match and patch Metal3DataTemplate or IPPool. Or am I missing something?

:disappointed: You are right. That is too bad.
Probably the most reasonable thing to do then is to keep them separate, like you say. So the user of the ClusterClass would have to create a Cluster, IPPool and Metal3DataTemplate. Doesn't sound too bad actually.

**Comment by p-strusiewiczsurmacki-mobica** [comment]
Moved `Metal3DataTemplates` and `IPPools` to `cluster.yaml` to keep those together.

**Review by lentzi90 — COMMENTED** [review]
/approve
This looks good to me, but you need to squash the commits before merging.
Adding hold so we don't forget it
/hold

**Comment by p-strusiewiczsurmacki-mobica** [comment]
@lentzi90 Just squashed the commits.

Also just found another issue - `Metal3DataTemplates` used by `Cluster` also have to be overridden using variables. If not, `ClusterClass` will always reference the same `Metal3DataTemplates` with fixed value of `clusterName`. So, I've just added variables `workerDataTemplate` and `controlPlaneDataTemplate` witch appropriate patches to the `ClusterClass`.

**Review by lentzi90 — COMMENTED** [review]
Thanks!
/hold cancel

**Comment by p-strusiewiczsurmacki-mobica** [comment]
Created PR with some explanation: https://github.com/metal3-io/metal3-docs/pull/385

**Comment by furkatgofurov7** [comment]
@p-strusiewiczsurmacki-mobica thanks for this work and documentation PR, really appreciated. 

The only thing I feel missing personally is having proper e2e tests to actually test this new feature. 
We have 2 mandatory CI jobs here to pass before we can take this PR, however, for this patch specifically, they should not matter because we can't really test this scenario. When introducing new features to Metal3, we usually follow with feature implementation and e2e test coverage (more about e2e package [here](https://github.com/metal3-io/cluster-api-provider-metal3/tree/main/test/e2e)) so that we have a way to test and make sure the feature is not broken in the future. 

If the scope of this PR was just to implement CC feature, then let's take this in and work on adding e2e tests as a follow-up (by creating a separate issue) otherwise we can wait until they are ready and run [e2e tests CI jobs](https://github.com/metal3-io/cluster-api-provider-metal3/tree/main?tab=readme-ov-file#feature-tests) which should give us a better confidence.

**Comment by p-strusiewiczsurmacki-mobica** [comment]
@furkatgofurov7 I've taken a quick look into the e2e tests and I think that some changes in `metal3-dev-env` repository will be required. Am I right?
I'll try to figure out how to add those tests, but I think that this could be a follow up task.

**Comment by p-strusiewiczsurmacki-mobica** [comment]
@furkatgofurov7 I've added some e2e tests. Currently only `GINKGO_FOCUS=basic` is supported as extended tests fail for me (but those also fail for me when using the main branch).

Tests can be run with `make test-clusterclass-e2e`.

File: config/default/capm3/manager_image_patch.yaml
Comment (furkatgofurov7) [inline config/default/capm3/manager_image_patch.yaml:11]: We can't change this and ipam image below to point to local images, please revert. You should be able to use something https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/Makefile#L454-L462 on the fly while running e2e tests but not set as is in upstream repo

File: config/ipam/kustomization.yaml
Comment (furkatgofurov7) [inline config/ipam/kustomization.yaml:6]: this is a problem, with original version we have a way to deploy ipam with specific version, usually it is aligned with CAPM3, but now we are hardcoding it. What are you trying to achieve with that approach?

**Review by furkatgofurov7 — COMMENTED** [review]
Thanks working on adding it, few comments inline:

File: config/default/capm3/manager_image_patch.yaml
Comment (p-strusiewiczsurmacki-mobica) [reply config/default/capm3/manager_image_patch.yaml]: I think those files in `/config` were changed when I've run the test and I have should not commit those. Reverted.
Sorry for that.

File: config/ipam/kustomization.yaml
Comment (p-strusiewiczsurmacki-mobica) [reply config/ipam/kustomization.yaml]: I think those files in `/config` were changed when I've run the test and I have should not commit those. Reverted.
Sorry for that.

File: test/e2e/data/infrastructure-metal3/bases/clusterclass-cluster/md.yaml
Comment (furkatgofurov7) [inline test/e2e/data/infrastructure-metal3/bases/clusterclass-cluster/md.yaml:1]: why is this commented out?

**Review by furkatgofurov7 — COMMENTED** [review]
This should be almost ready

File: test/e2e/test1-cluster-template.yaml
Comment (p-strusiewiczsurmacki-mobica) [reply test/e2e/test1-cluster-template.yaml]: Deleted this.

**Review by furkatgofurov7 — APPROVED** [review]
Thanks for pushing this forward @p-strusiewiczsurmacki-mobica. There is a small nit, otherwise it looks good.

File: Makefile
Comment (furkatgofurov7) [reply Makefile:175]: @p-strusiewiczsurmacki-mobica hey, will you have a time to address this comment soon? We can then try to run CI tests on this one.

Comment (p-strusiewiczsurmacki-mobica) [reply Makefile:175]: @furkatgofurov7 Sorry, I've had to omit the notification email for this comment, and I've only seen it now.

I think that this should stay here. `clusterclass-templates` will not be used if non-clusterclass e2e test is being run, and `clusterclass.yaml` file is required by [e2e config](https://github.com/p-strusiewiczsurmacki-mobica/cluster-api-provider-metal3/blob/clusterclass-support/test/e2e/config/e2e_conf.yaml#L106). If it will not exist, the test will fail.
I had no better idea on how to fix that, but to create empty `clusterclass.yaml` file.

Comment (furkatgofurov7) [reply Makefile:175]: ah right, now I got the point. Ideally, we could test the feature (simple cluster creation/destroy using CC would have been enough) as part of the current e2e tests setup, instead of having a separate Makefile target which is not part of it right now. 
@lentzi90 WDYT of the current approach?

Comment (lentzi90) [reply Makefile:175]: I think we can live with it, but I agree that making it part of the current e2e tests would be preferable. Enabling cluster class support in the controllers should be fine for all tests. They will anyway not use a ClusterClass unless defined in the test :slightly_smiling_face:

**Comment by kashifest** [comment]
This would need squashing of the commits I suppose, @furkatgofurov7 PTAL

**Comment by p-strusiewiczsurmacki-mobica** [comment]
@furkatgofurov7 

> And yes, the last 2 commits could be squashed into 1, preferably.

I've left 2 commits now. Is it OK, or should squash it into 1?

Also, I've just rebased the branch as there were some conflicts. Trying to run e2e tests to check if everything is correct just now.

**Comment by p-strusiewiczsurmacki-mobica** [comment]
Well, for now e2e tests both on this branch and `main` branch fails for me with:

```
TASK [libvirt : Start libvirt networks] ****************************************
changed: [localhost] => (item={'name': 'provisioning', 'bridge': 'provisioning', 'forward_mode': 'bridge'})
failed: [localhost] (item={'name': 'external', 'bridge': 'external', 'forward_mode': 'nat', 'address_v4': '192.168.111.1', 'netmask_v4': '255.255.255.0', 'dhcp_range_v4': ['192.168.111.20', '192.168.111.60'], 'address_v6': '', 'prefix_v6': False, 'dhcp_range_v6': ['', ''], 'lease_expiry': 60, 'nat_port_range': [1024, 65535], 'domain': 'ostest.test.metalkube.org', 'dns': {'hosts': [], 'forwarders': [{'domain': 'apps.ostest.test.metalkube.org', 'addr': '127.0.0.1'}], 'srvs': []}}) => {"ansible_loop_var": "item", "changed": false, "item": {"address_v4": "192.168.111.1", "address_v6": "", "bridge": "external", "dhcp_range_v4": ["192.168.111.20", "192.168.111.60"], "dhcp_range_v6": ["", ""], "dns": {"forwarders": [{"addr": "127.0.0.1", "domain": "apps.ostest.test.metalkube.org"}], "hosts": [], "srvs": []}, "domain": "ostest.test.metalkube.org", "forward_mode": "nat", "lease_expiry": 60, "name": "external", "nat_port_range": [1024, 65535], "netmask_v4": "255.255.255.0", "prefix_v6": false}, "msg": "internal error: Child process (VIR_BRIDGE_NAME=external /usr/sbin/dnsmasq --conf-file=/var/lib/libvirt/dnsmasq/external.conf --leasefile-ro --dhcp-script=/usr/lib/libvirt/libvirt_leaseshelper) unexpected exit status 2: \ndnsmasq: failed to create listening socket for 192.168.111.1: Address already in use\n"}
```

Something inside the test starts listing on
```
tcp        0      0 192.168.111.1:53        0.0.0.0:*               LISTEN      -
udp        0      0 192.168.111.1:53        0.0.0.0:*                           -
```

Nothing is listening on this address before I start the test.

**Comment by p-strusiewiczsurmacki-mobica** [comment]
Well, both this branch and `main` branch just kills my VM when first node is being provisioned. Need to investigate that.

**Comment by furkatgofurov7** [comment]
>I've left 2 commits now. Is it OK, or should squash it into 1?

I am fine with as is (2 commits).

**Comment by furkatgofurov7** [comment]
@kashifest tests are not triggered, is this a know issue or triggers have changed?

**Comment by LingyanCao** [comment]
May I know when could this PR be merged and when is the next release that contains this change?

**Comment by kashifest** [comment]
> May I know when could this PR be merged and when is the next release that contains this change? @p-strusiewiczsurmacki-mobica Our product has the ClusterClass dependency of capm3. I want to raise this risk if there is schedule. Thank you in advance!

I think this is ready to go in , pending a rebase

File: examples/templates/clusterclass.yaml
Comment (chess-knight) [inline examples/templates/clusterclass.yaml:57]: This doesn't work for me(virtualized env). Shouldn't it be `'{{ ds.meta_data.local_hostname }}'` or `'{{ ds.meta_data.name }}'`?. The question is also, what is the difference between `local_hostname` and `name` because I see both here.

**Comment by p-strusiewiczsurmacki-mobica** [comment]
> > May I know when could this PR be merged and when is the next release that contains this change? @p-strusiewiczsurmacki-mobica Our product has the ClusterClass dependency of capm3. I want to raise this risk if there is schedule. Thank you in advance!
> 
> I think this is ready to go in , pending a rebase

Hi, I've just rebased it :)

File: examples/templates/clusterclass.yaml
Comment (kashifest) [reply examples/templates/clusterclass.yaml:57]: We can fix it in a followup or leave it out completely from the template in a followup PR. @p-strusiewiczsurmacki-mobica would you be interested in a followup PR? For now, we take this PR in

Comment (p-strusiewiczsurmacki-mobica) [reply examples/templates/clusterclass.yaml:57]: @kashifest I'll try to take a look at this as soon as I can. However if anyone wants to work on this before I'll be able to, then go ahead. :)

**Pull request merged**
